Core/Core/Use Cases/UploadBatch.swift

```diff
@@ -0,0 +1,142 @@
+//
+// Copyright (C) 2019-present Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, version 3 of the License.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+import Foundation
+
+public class UploadBatch {
+    public typealias Callback = (State?) -> Void
+
+    public enum State: Equatable {
+        case staged
+        case uploading
+        case failed(Error)
+        case completed(fileIDs: Set<String>)
+
+        public static func == (lhs: State, rhs: State) -> Bool {
+            switch (lhs, rhs) {
+            case (.staged, .staged): return true
+            case (.uploading, .uploading): return true
+            case (.failed, .failed): return true
+            case let (.completed(fileIDs: lhs), .completed(fileIDs: rhs)): return lhs == rhs
+            default: return false
+            }
+        }
```

Comment:
I'm surprised we would need a custom implementation. Usually for these kinds of enums, simply declaring `Equatable` is enough.

Reply:
Yeah I think the associated types throw it off.

Reply:
`APIQueryItem`, `APIFormDatum`, `Login`, `RouteHandler.Segment` all have associated types,  are `Equatable`, and don't require a custom implementation.

Reply:
Oh interesting. I figured it out. It's the `case failed(Error)` that breaks it since `Error` isn't equatable.

Reply:
ok, that makes sense.